controllers/lock_client.go

```diff
@@ -233,6 +248,79 @@ func (client *RealLockClient) ClearPendingUpgrades() error {
 	return err
 }
 
+// GetDenyList retrieves the current deny list from the database.
+func (client *RealLockClient) GetDenyList() ([]string, error) {
+	list, err := client.database.Transact(func(tr fdb.Transaction) (interface{}, error) {
+		err := tr.Options().SetReadSystemKeys()
+		if err != nil {
+			return nil, err
+		}
+
+		keyRange, err := client.getDenyListKeyRange()
+		if err != nil {
+			return nil, err
+		}
+
+		values := tr.GetRange(keyRange, fdb.RangeOptions{}).GetSliceOrPanic()
+		list := make([]string, len(values))
+		for index, value := range values {
+			list[index] = string(value.Value)
+		}
+		return list, nil
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	return list.([]string), nil
+}
+
+// UpdateDenyList updates the deny list to match a list of entries.
+// This will return the complete deny list after these changes are made.
+func (client *RealLockClient) UpdateDenyList(locks []fdbtypes.LockDenyListEntry) error {
+	_, err := client.database.Transact(func(tr fdb.Transaction) (interface{}, error) {
+		err := tr.Options().SetAccessSystemKeys()
+		if err != nil {
+			return nil, err
+		}
+
+		keyRange, err := client.getDenyListKeyRange()
+		if err != nil {
+			return nil, err
+		}
+
+		values := tr.GetRange(keyRange, fdb.RangeOptions{}).GetSliceOrPanic()
+		denyListMap := make(map[string]bool, len(values))
+		for _, value := range values {
+			denyListMap[string(value.Value)] = true
+		}
+
+		for _, entry := range locks {
+			if entry.Allow && denyListMap[entry.ID] {
```

Comment:
What should happen if `entry.Allow` is `false` and the ID is set in the `denyListMap`? I would assume that we want to add the ID to the deny list?

I also would expect that `locks` represents the current expected state of the system so do we really need to check the current state or should we just set the expected state from `locks`?

Reply:
The `locks` field can be set differently in different KCs, so we need to allow for keeping things in the deny list that a given instance of the operator doesn't have any deny list configuration for.

If `entry.Allow` is false and the ID is set in the `denyListMap`, then the deny list is correctly configured and we don't need to make changes here.